# Segmenting and Clustering Neighbourhoods in Toronto - Peer Graded Assignment

### Note: All the Three parts (Questions 1, 2 & 3) of the Peer Graded Assignment are included in this Notebook

In [1]:
# Importing Libraries
import pandas as pd
import numpy as np

import requests

from bs4 import BeautifulSoup

In [2]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"

In [3]:
wiki = requests.get(url).text
wiki

'<!DOCTYPE html>\n<html class="client-nojs" lang="en" dir="ltr">\n<head>\n<meta charset="UTF-8"/>\n<title>List of postal codes of Canada: M - Wikipedia</title>\n<script>document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"da975d4f-cf55-4923-bf5a-f430002c74d9","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":1032600019,"wgRevisionId":1032600019,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Short description is different from Wikidata","Communications in Ont

## Question 1:

#### Use Pandas or BeautifulSoup or any other way you are comfortable with to transform the data in the table on the Wikipedia page into the above pandas dataframe.

In [4]:
soup = BeautifulSoup(wiki, "html.parser")
print(soup.prettify())

<!DOCTYPE html>
<html class="client-nojs" dir="ltr" lang="en">
 <head>
  <meta charset="utf-8"/>
  <title>
   List of postal codes of Canada: M - Wikipedia
  </title>
  <script>
   document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"da975d4f-cf55-4923-bf5a-f430002c74d9","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":1032600019,"wgRevisionId":1032600019,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Short description is different from Wikidata","Communica

In [5]:
wiki_tables = soup.find("table")
wiki_tables

<table cellpadding="2" cellspacing="0" rules="all" style="width:100%;">
<tbody><tr>
<td style="width:11%;">
<p>M1A<br/><span style="font-size:85%;">Not assigned</span>
</p>
</td>
<td style="width:11%;">
<p>M2A<br/><span style="font-size:85%;">Not assigned</span>
</p>
</td>
<td style="width:11%;">
<p>M3A<br/><span style="font-size:85%;"><a href="/wiki/North_York" title="North York">North York</a><br/>(<a href="/wiki/Parkwoods" title="Parkwoods">Parkwoods</a>)</span>
</p>
</td>
<td style="width:11%;">
<p>M4A<br/><span style="font-size:85%;"><a href="/wiki/North_York" title="North York">North York</a><br/>(<a href="/wiki/Victoria_Village" title="Victoria Village">Victoria Village</a>)</span>
</p>
</td>
<td style="width:11%;">
<p>M5A<br/><span style="font-size:85%;"><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a><br/>(<a href="/wiki/Regent_Park" title="Regent Park">Regent Park</a> / <a href="/wiki/Harbourfront,_Toronto" title="Harbourfront, Toronto">Harbourfr

In [6]:
# Create an empty Database with columns
column_names = ["PostalCode", "Borough", "Neighborhood"]
df = pd.DataFrame(columns = column_names)

In [7]:
table_contents=[]

for row in wiki_tables.findAll("td"):
    cell = {}
    if row.span.text == "Not assigned":         # Excluding the records where Neighborhoord are "Not Assigned"
        pass
    else:
        cell["PostalCode"] = row.p.text[ : 3]
        cell["Borough"] = (row.span.text).split("(")[0]
        cell["Neighborhood"] = (((((row.span.text).split("(")[1]).strip(")")).replace(" /", ",")).replace(")", " ")).strip(" ")
        table_contents.append(cell)

print(table_contents)

[{'PostalCode': 'M3A', 'Borough': 'North York', 'Neighborhood': 'Parkwoods'}, {'PostalCode': 'M4A', 'Borough': 'North York', 'Neighborhood': 'Victoria Village'}, {'PostalCode': 'M5A', 'Borough': 'Downtown Toronto', 'Neighborhood': 'Regent Park, Harbourfront'}, {'PostalCode': 'M6A', 'Borough': 'North York', 'Neighborhood': 'Lawrence Manor, Lawrence Heights'}, {'PostalCode': 'M7A', 'Borough': "Queen's Park", 'Neighborhood': 'Ontario Provincial Government'}, {'PostalCode': 'M9A', 'Borough': 'Etobicoke', 'Neighborhood': 'Islington Avenue'}, {'PostalCode': 'M1B', 'Borough': 'Scarborough', 'Neighborhood': 'Malvern, Rouge'}, {'PostalCode': 'M3B', 'Borough': 'North York', 'Neighborhood': 'Don Mills North'}, {'PostalCode': 'M4B', 'Borough': 'East York', 'Neighborhood': 'Parkview Hill, Woodbine Gardens'}, {'PostalCode': 'M5B', 'Borough': 'Downtown Toronto', 'Neighborhood': 'Garden District, Ryerson'}, {'PostalCode': 'M6B', 'Borough': 'North York', 'Neighborhood': 'Glencairn'}, {'PostalCode': 'M9

In [8]:
df = pd.DataFrame(table_contents)
df.head()

PostalCode           Borough                      Neighborhood
0        M3A        North York                         Parkwoods
1        M4A        North York                  Victoria Village
2        M5A  Downtown Toronto         Regent Park, Harbourfront
3        M6A        North York  Lawrence Manor, Lawrence Heights
4        M7A      Queen's Park     Ontario Provincial Government

In [9]:
df = df.groupby(["PostalCode", "Borough"], sort = True).agg(", ".join)
df.reset_index(inplace = True)

In [10]:
df

PostalCode             Borough  \
0          M1B         Scarborough   
1          M1C         Scarborough   
2          M1E         Scarborough   
3          M1G         Scarborough   
4          M1H         Scarborough   
..         ...                 ...   
98         M9N                York   
99         M9P           Etobicoke   
100        M9R           Etobicoke   
101        M9V           Etobicoke   
102        M9W  EtobicokeNorthwest   

                                          Neighborhood  
0                                       Malvern, Rouge  
1               Rouge Hill, Port Union, Highland Creek  
2                    Guildwood, Morningside, West Hill  
3                                               Woburn  
4                                            Cedarbrae  
..                                                 ...  
98                                              Weston  
99                                           Westmount  
100  Kingsview Village, St. Phillips, Martin Grove ...  
101  South Steeles, Silverstone, Humbergate, Jamest...  
102  Clairville, Humberwood, Woodbine Downs, West H...  

[103 rows x 3 columns]

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103 entries, 0 to 102
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   PostalCode    103 non-null    object
 1   Borough       103 non-null    object
 2   Neighborhood  103 non-null    object
dtypes: object(3)
memory usage: 2.5+ KB


In [12]:
df.shape

(103, 3)

### The dataframe has 103 rows and 3 columns 

## Question 2:

In [13]:
import geocoder

### Interupted the following code after running it for more than 2 hours. Commenting the following code so as not to hanf the system

In [14]:
# initialize your variable to None
#lat_lng_coords = None

# initialize the variable to "M3A"
#postal_code = "M3A"

# loop until you get the coordinates
#while(lat_lng_coords is None):
#  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
#  lat_lng_coords = g.latlng

#latitude = lat_lng_coords[0]
#longitude = lat_lng_coords[1]

### SInce I am not able to get the geographical coordinates of the neighborhoods using the Geocoder package, using the below mentioned link to the csv file that has the geographical coordinates of each postal code as suggested in the assignment

In [15]:
dataset = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs_v1/Geospatial_Coordinates.csv"

In [16]:
dataset

'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs_v1/Geospatial_Coordinates.csv'

In [17]:
geospatial_df = pd.read_csv(dataset)

In [18]:
geospatial_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [19]:
geospatial_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103 entries, 0 to 102
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Postal Code  103 non-null    object 
 1   Latitude     103 non-null    float64
 2   Longitude    103 non-null    float64
dtypes: float64(2), object(1)
memory usage: 2.5+ KB


### Since the number of rows in the "geospatial_df" and the "df" dataframes match, we can join both the dataframes using the "Postal Code" column.

In [20]:
# Change the Column Name of "PostalCode" in "df" dataframe to "Postal Code" to match that with the "geospatial_df" dataframe
df.rename(columns = {"PostalCode" : "Postal Code"}, inplace = True)
df.head()

Postal Code      Borough                            Neighborhood
0         M1B  Scarborough                          Malvern, Rouge
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2         M1E  Scarborough       Guildwood, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                               Cedarbrae

In [21]:
# df_new = df.join(geospatial_df.set_index("Postal Code"), on = "Postal Code", how = "inner")
# OR
df_new = pd.merge(df, geospatial_df, on = "Postal Code")
df_new

Postal Code             Borough  \
0           M1B         Scarborough   
1           M1C         Scarborough   
2           M1E         Scarborough   
3           M1G         Scarborough   
4           M1H         Scarborough   
..          ...                 ...   
98          M9N                York   
99          M9P           Etobicoke   
100         M9R           Etobicoke   
101         M9V           Etobicoke   
102         M9W  EtobicokeNorthwest   

                                          Neighborhood   Latitude  Longitude  
0                                       Malvern, Rouge  43.806686 -79.194353  
1               Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497  
2                    Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.239476  
..                                                 ...        ...        ...  
98                                              Weston  43.706876 -79.518188  
99                                           Westmount  43.696319 -79.532242  
100  Kingsview Village, St. Phillips, Martin Grove ...  43.688905 -79.554724  
101  South Steeles, Silverstone, Humbergate, Jamest...  43.739416 -79.588437  
102  Clairville, Humberwood, Woodbine Downs, West H...  43.706748 -79.594054  

[103 rows x 5 columns]

In [22]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 103 entries, 0 to 102
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Postal Code   103 non-null    object 
 1   Borough       103 non-null    object 
 2   Neighborhood  103 non-null    object 
 3   Latitude      103 non-null    float64
 4   Longitude     103 non-null    float64
dtypes: float64(2), object(3)
memory usage: 4.8+ KB


In [23]:
df_new.shape

(103, 5)

## Question 3:

### Creating a new Dataframe with Boroughs that contain the word Toronto

In [24]:
df_toronto = df_new[df_new["Borough"].str.contains("Toronto", regex = False)]
df_toronto.head()

Postal Code                Borough                    Neighborhood  \
37         M4E           East Toronto                     The Beaches   
40         M4J  East YorkEast Toronto              The Danforth  East   
41         M4K           East Toronto    The Danforth West, Riverdale   
42         M4L           East Toronto  India Bazaar, The Beaches West   
43         M4M           East Toronto                 Studio District   

     Latitude  Longitude  
37  43.676357 -79.293031  
40  43.685347 -79.338106  
41  43.679557 -79.352188  
42  43.668999 -79.315572  
43  43.659526 -79.340923

In [25]:
df_toronto.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 39 entries, 37 to 87
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Postal Code   39 non-null     object 
 1   Borough       39 non-null     object 
 2   Neighborhood  39 non-null     object 
 3   Latitude      39 non-null     float64
 4   Longitude     39 non-null     float64
dtypes: float64(2), object(3)
memory usage: 1.8+ KB


In [26]:
import folium
import requests 
import json 
import matplotlib.cm as cm
import matplotlib.colors as colors
from pandas.io.json import json_normalize 
from sklearn.cluster import KMeans
from geopy.geocoders import Nominatim

In [27]:
address = "Toronto, Ontario"

geolocator = Nominatim(user_agent = "toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


### Map of Toronto Using the Entire Dataframe

In [28]:
# Creating the map of Toronto
map_Toronto_dfnew = folium.Map(location = [latitude, longitude], zoom_start = 11)

# Adding markers to map
for latitude, longitude, borough, neighborhood in zip(df_new['Latitude'], df_new['Longitude'], df_new['Borough'], df_new['Neighborhood']):
    label = "{}, {}".format(neighborhood, borough)
    label = folium.Popup(label, parse_html = True)
    folium.CircleMarker(
                            [latitude, longitude],
                            radius = 5,
                            popup = label,
                            color = "blue",
                            fill = True
                        ).add_to(map_Toronto_dfnew)  
    
map_Toronto_dfnew

### Map of Toronto Using the Toronto Dataframe

In [29]:
# Creating the map of Toronto
map_Toronto = folium.Map(location = [latitude, longitude], zoom_start = 11)

# Adding markers to map
for latitude, longitude, borough, neighborhood in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Borough'], df_toronto['Neighborhood']):
    label = "{}, {}".format(neighborhood, borough)
    label = folium.Popup(label, parse_html = True)
    folium.CircleMarker(
                            [latitude, longitude],
                            radius = 5,
                            popup = label,
                            color = "red",
                            fill = True
                        ).add_to(map_Toronto)  
    
map_Toronto

### Define Foursquare API Credentials

In [30]:
CLIENT_ID = "DVS0H5U52KP03MEJUX0WCA0SZ1SQB2ASHNNRRB4SIT2GWJ3U" # your Foursquare ID
CLIENT_SECRET = "SCZJLODKBUVU2BSHU3MAEKUXR0BRDIDCQILD3QQW1X4PUW2Y" # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: DVS0H5U52KP03MEJUX0WCA0SZ1SQB2ASHNNRRB4SIT2GWJ3U
CLIENT_SECRET:SCZJLODKBUVU2BSHU3MAEKUXR0BRDIDCQILD3QQW1X4PUW2Y


### Create a Function to Get All the Venue Categories in Toronto

In [31]:
def getNearbyVenues(names, latitudes, longitudes, radius = 500):
    
    venues_list = []
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = "https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}".format(
                    CLIENT_ID, 
                    CLIENT_SECRET, 
                    VERSION, 
                    lat, 
                    lng, 
                    radius
                    )
            
        # make the GET request
        results = requests.get(url).json()["response"]["groups"][0]["items"]
        
        # return only relevant information for each nearby venue
        venues_list.append([(
                                name, 
                                lat, 
                                lng, 
                                v["venue"]["name"],
                                v["venue"]["location"]["lat"],
                                v["venue"]["location"]["lng"],
                                v["venue"]["categories"][0]["name"]
                            ) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ["Neighborhood", "Neighborhood Latitude", "Neighborhood Longitude", "Venue", "Venue Latitude", "Venue Longitude", "Venue Category"]
    
    return (nearby_venues)

### Collecting the venues in Toronto for each Neighbourhood Using the Entire Dataframe

In [32]:
venues_in_toronto_dfnew = getNearbyVenues(df_new["Neighborhood"], df_new["Latitude"], df_new["Longitude"])

Malvern, Rouge
Rouge Hill, Port Union, Highland Creek
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
Kennedy Park, Ionview, East Birchmount Park
Golden Mile, Clairlea, Oakridge
Cliffside, Cliffcrest, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Wexford Heights, Scarborough Town Centre
Wexford, Maryvale
Agincourt
Clarks Corners, Tam O'Shanter, Sullivan
Milliken, Agincourt North, Steeles East, L'Amoreaux East
Steeles West, L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
York Mills, Silver Hills
Willowdale, Newtonbrook
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Don Mills South
Bathurst Manor, Wilson Heights, Downsview North
Northwood Park, York University
Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Parkview Hill, Woodbine Gardens
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
The Danforth  East
The Danforth West, Riverdale


In [33]:
venues_in_toronto_dfnew.shape

(1331, 7)

In [34]:
venues_in_toronto_dfnew.head()

Neighborhood  Neighborhood Latitude  \
0                          Malvern, Rouge              43.806686   
1  Rouge Hill, Port Union, Highland Creek              43.784535   
2       Guildwood, Morningside, West Hill              43.763573   
3       Guildwood, Morningside, West Hill              43.763573   
4       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                  Venue  Venue Latitude  \
0              -79.194353                Wendy’s       43.807448   
1              -79.160497  Royal Canadian Legion       43.782533   
2              -79.188711         RBC Royal Bank       43.766790   
3              -79.188711      G & G Electronics       43.765309   
4              -79.188711             Sail Sushi       43.765951   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.163085                   Bar  
2       -79.191151                  Bank  
3       -79.191537     Electronics Store  
4       -79.191275            Restaurant

### Collecting the venues in Toronto for each Neighbourhood Using the Toronto Dataframe

In [35]:
venues_in_toronto = getNearbyVenues(df_toronto["Neighborhood"], df_toronto["Latitude"], df_toronto["Longitude"])

The Beaches
The Danforth  East
The Danforth West, Riverdale
India Bazaar, The Beaches West
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park, Summerhill East
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North & West
The Annex, North Midtown, Yorkville
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Enclave of M5E
First Canadian Place, Underground city
Christie
Dufferin, Dovercourt Village
Little Portugal, Trinity
Brockton, Parkdale Village, Exhibition Place
High Park, The Junction 

In [36]:
venues_in_toronto.shape

(835, 7)

In [37]:
venues_in_toronto.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         The Beaches              43.676357              -79.293031   
1         The Beaches              43.676357              -79.293031   
2         The Beaches              43.676357              -79.293031   
3         The Beaches              43.676357              -79.293031   
4  The Danforth  East              43.685347              -79.338106   

                                Venue  Venue Latitude  Venue Longitude  \
0                   Glen Manor Ravine       43.676821       -79.293942   
1  The Big Carrot Natural Food Market       43.678879       -79.297734   
2                 Grover Pub and Grub       43.679181       -79.297215   
3                       Upper Beaches       43.680563       -79.292869   
4                            The Path       43.683923       -79.335007   

      Venue Category  
0              Trail  
1  Health Food Store  
2                Pub  
3       Neighborhood  
4               Park

### Checking the Venues based on Neighbourhood Using the Entire Dataframe

In [38]:
venues_in_toronto_dfnew.groupby('Neighborhood').head()

Neighborhood  \
0                                        Malvern, Rouge   
1                Rouge Hill, Port Union, Highland Creek   
2                     Guildwood, Morningside, West Hill   
3                     Guildwood, Morningside, West Hill   
4                     Guildwood, Morningside, West Hill   
...                                                 ...   
1322  South Steeles, Silverstone, Humbergate, Jamest...   
1327  Clairville, Humberwood, Woodbine Downs, West H...   
1328  Clairville, Humberwood, Woodbine Downs, West H...   
1329  Clairville, Humberwood, Woodbine Downs, West H...   
1330  Clairville, Humberwood, Woodbine Downs, West H...   

      Neighborhood Latitude  Neighborhood Longitude                   Venue  \
0                 43.806686              -79.194353                 Wendy’s   
1                 43.784535              -79.160497   Royal Canadian Legion   
2                 43.763573              -79.188711          RBC Royal Bank   
3                 43.763573              -79.188711       G & G Electronics   
4                 43.763573              -79.188711              Sail Sushi   
...                     ...                     ...                     ...   
1322              43.739416              -79.588437     Sheriff's No Frills   
1327              43.706748              -79.594054      Economy Rent A Car   
1328              43.706748              -79.594054  Logistics Distribution   
1329              43.706748              -79.594054           Saand Rexdale   
1330              43.706748              -79.594054    Vectra Heavy Haulers   

      Venue Latitude  Venue Longitude        Venue Category  
0          43.807448       -79.199056  Fast Food Restaurant  
1          43.782533       -79.163085                   Bar  
2          43.766790       -79.191151                  Bank  
3          43.765309       -79.191537     Electronics Store  
4          43.765951       -79.191275            Restaurant  
...              ...              ...                   ...  
1322       43.741696       -79.584379         Grocery Store  
1327       43.708471       -79.589943   Rental Car Location  
1328       43.707554       -79.589252                   Bar  
1329       43.705072       -79.598725             Drugstore  
1330       43.704891       -79.599410            Truck Stop  

[428 rows x 7 columns]

#### There are 428 records for each neighbourhood using the entire dataframe

In [39]:
venues_in_toronto_dfnew.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                             
Agincourt                                                            4   
Alderwood, Long Branch                                               8   
Bathurst Manor, Wilson Heights, Downsview North                     21   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   25   
...                                                                ...   
Willowdale West                                                      5   
Willowdale, Newtonbrook                                              1   
Woburn                                                               4   
Woodbine Heights                                                     8   
York Mills West                                                      3   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Agincourt                                                             4   
Alderwood, Long Branch                                                8   
Bathurst Manor, Wilson Heights, Downsview North                      21   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    25   
...                                                                 ...   
Willowdale West                                                       5   
Willowdale, Newtonbrook                                               1   
Woburn                                                                4   
Woodbine Heights                                                      8   
York Mills West                                                       3   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Agincourt                                            4               4   
Alderwood, Long Branch                               8               8   
Bathurst Manor, Wilson Heights, Downsview North     21              21   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   25              25   
...                                                ...             ...   
Willowdale West                                      5               5   
Willowdale, Newtonbrook                              1               1   
Woburn                                               4               4   
Woodbine Heights                                     8               8   
York Mills West                                      3               3   

                                                 Venue Longitude  \
Neighborhood                                                       
Agincourt                                                      4   
Alderwood, Long Branch                                         8   
Bathurst Manor, Wilson Heights, Downsview North               21   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             25   
...                                                          ...   
Willowdale West                                                5   
Willowdale, Newtonbrook                                        1   
Woburn                                                         4   
Woodbine Heights                                               8   
York Mills West                                                3   

                                                 Venue Category  
Neighborhood                                                     
Agincourt                                                     4  
Alderwood, Long Branch                                        8 

### Checking the Venues based on Neighbourhood Using the Toronto Dataframe

In [40]:
venues_in_toronto.groupby('Neighborhood').head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0           The Beaches              43.676357              -79.293031   
1           The Beaches              43.676357              -79.293031   
2           The Beaches              43.676357              -79.293031   
3           The Beaches              43.676357              -79.293031   
4    The Danforth  East              43.685347              -79.338106   
..                  ...                    ...                     ...   
819      Enclave of M4L              43.662744              -79.321558   
820      Enclave of M4L              43.662744              -79.321558   
821      Enclave of M4L              43.662744              -79.321558   
822      Enclave of M4L              43.662744              -79.321558   
823      Enclave of M4L              43.662744              -79.321558   

                                  Venue  Venue Latitude  Venue Longitude  \
0                     Glen Manor Ravine       43.676821       -79.293942   
1    The Big Carrot Natural Food Market       43.678879       -79.297734   
2                   Grover Pub and Grub       43.679181       -79.297215   
3                         Upper Beaches       43.680563       -79.292869   
4                              The Path       43.683923       -79.335007   
..                                  ...             ...              ...   
819               Rorschach Brewing Co.       43.663483       -79.319824   
820          Leslieville Farmers Market       43.664901       -79.319784   
821                        The Sidekick       43.664484       -79.325162   
822              Queen Margherita Pizza       43.664685       -79.324164   
823                         Chino Locos       43.664653       -79.325584   

        Venue Category  
0                Trail  
1    Health Food Store  
2                  Pub  
3         Neighborhood  
4                 Park  
..                 ...  
819            Brewery  
820     Farmers Market  
821         Comic Shop  
822        Pizza Place  
823      Burrito Place  

[186 rows x 7 columns]

#### There are 186 records for each neighbourhood using the Toronto dataframe

In [41]:
venues_in_toronto.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            30   
Brockton, Parkdale Village, Exhibition Place                           23   
CN Tower, King and Spadina, Railway Lands, Harb...                     14   
Central Bay Street                                                     30   
Christie                                                               16   
Church and Wellesley                                                   30   
Commerce Court, Victoria Hotel                                         30   
Davisville                                                             30   
Davisville North                                                        9   
Dufferin, Dovercourt Village                                           16   
Enclave of M4L                                                         16   
Enclave of M5E                                                         30   
First Canadian Place, Underground city                                 30   
Forest Hill North & West                                                4   
Garden District, Ryerson                                               30   
Harbourfront East, Union Station, Toronto Islands                      30   
High Park, The Junction South                                          25   
India Bazaar, The Beaches West                                         19   
Kensington Market, Chinatown, Grange Park                              30   
Lawrence Park                                                           5   
Little Portugal, Trinity                                               30   
Moore Park, Summerhill East                                             3   
North Toronto West                                                     22   
Parkdale, Roncesvalles                                                 14   
Regent Park, Harbourfront                                              30   
Richmond, Adelaide, King                                               30   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede, Swansea                                                     30   
St. James Town                                                         30   
St. James Town, Cabbagetown                                            30   
Studio District                                                        30   
Summerhill West, Rathnelly, South Hill, Forest ...                     14   
The Annex, North Midtown, Yorkville                                    21   
The Beaches                                                             4   
The Danforth  East                                                      4   
The Danforth West, Riverdale                                           30   
Toronto Dominion Centre, Design Exchange                               30   
University of Toronto, Harbord                                         30   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             30   
Brockton, Parkdale Village, Exhibition Place                            23   
CN Tower, King and Spadina, Railway Lands, Harb...                      14   
Central Bay Street                                                      30   
Christie                                                                16   
Church and Wellesley                                                    30   
Commerce Court, Victoria Hotel                                          30   
Davisville                                                              30   
Davisville North                                                         9   
Dufferin, Dovercourt Village        

### Checking the Venue Categories Using the Entire Dataframe

In [42]:
venues_in_toronto_dfnew.groupby("Venue Category").max()

Neighborhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Adult Boutique                                   Church and Wellesley   
Airport                                                Downsview East   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                                     Little Portugal, Trinity   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                                     Caledonia-Fairbanks   
Yoga Studio                            University of Toronto, Harbord   

                    Neighborhood Latitude  Neighborhood Longitude  \
Venue Category                                                      
Accessories Store               43.718518              -79.464763   
Adult Boutique                  43.665860              -79.383160   
Airport                         43.737473              -79.394420   
Airport Food Court              43.628947              -79.394420   
Airport Lounge                  43.628947              -79.394420   
...                                   ...                     ...   
Warehouse Store                 43.705369              -79.349372   
Wine Bar                        43.653206              -79.400049   
Wings Joint                     43.628841              -79.520999   
Women's Store                   43.689026              -79.453512   
Yoga Studio                     43.715383              -79.321558   

                                              Venue  Venue Latitude  \
Venue Category                                                        
Accessories Store               Ardene Shoes Outlet       43.718892   
Adult Boutique                            Seduction       43.665620   
Airport             Toronto Downsview Airport (YZD)       43.738883   
Airport Food Court                Billy Bishop Café       43.631132   
Airport Lounge                        Porter Lounge       43.631360   
...                                             ...             ...   
Warehouse Store                              Costco       43.707051   
Wine Bar                            Paris Paris Bar       43.653479   
Wings Joint                              Wingporium       43.630275   
Women's Store                         Maximum Woman       43.690651   
Yoga Studio                               YogaSpace       43.714070   

                    Venue Longitude  
Venue Category                       
Accessories Store        -79.461344  
Adult Boutique           -79.384681  
Airport                  -79.396033  
Airport Food Court       -79.396139  
Airport Lounge           -79.395756  
...                             ...  
Warehouse Store          -79.348093  
Wine Bar                 -79.401427  
Wings Joint              -79.518169  
Women's Store            -79.456333  
Yoga Studio              -79.324335  

[240 rows x 6 columns]

#### There are 240 Venue Categories

In [43]:
venues_in_toronto_dfnew.groupby("Venue Category").count()

Neighborhood  Neighborhood Latitude  \
Venue Category                                            
Accessories Store              1                      1   
Adult Boutique                 1                      1   
Airport                        2                      2   
Airport Food Court             1                      1   
Airport Lounge                 2                      2   
...                          ...                    ...   
Warehouse Store                1                      1   
Wine Bar                       2                      2   
Wings Joint                    1                      1   
Women's Store                  1                      1   
Yoga Studio                   11                     11   

                    Neighborhood Longitude  Venue  Venue Latitude  \
Venue Category                                                      
Accessories Store                        1      1               1   
Adult Boutique                           1      1               1   
Airport                                  2      2               2   
Airport Food Court                       1      1               1   
Airport Lounge                           2      2               2   
...                                    ...    ...             ...   
Warehouse Store                          1      1               1   
Wine Bar                                 2      2               2   
Wings Joint                              1      1               1   
Women's Store                            1      1               1   
Yoga Studio                             11     11              11   

                    Venue Longitude  
Venue Category                       
Accessories Store                 1  
Adult Boutique                    1  
Airport                           2  
Airport Food Court                1  
Airport Lounge                    2  
...                             ...  
Warehouse Store                   1  
Wine Bar                          2  
Wings Joint                       1  
Women's Store                     1  
Yoga Studio                      11  

[240 rows x 6 columns]

### Checking the Venue Categories Using the Toronto Dataframe

In [44]:
venues_in_toronto.groupby("Venue Category").max()

Neighborhood  \
Venue Category                                                                     
Adult Boutique                                              Church and Wellesley   
Airport                        CN Tower, King and Spadina, Railway Lands, Har...   
Airport Food Court             CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge                 CN Tower, King and Spadina, Railway Lands, Har...   
Airport Service                CN Tower, King and Spadina, Railway Lands, Har...   
...                                                                          ...   
Vegetarian / Vegan Restaurant                                     St. James Town   
Video Game Store                                  University of Toronto, Harbord   
Vietnamese Restaurant          Summerhill West, Rathnelly, South Hill, Forest...   
Wine Bar                                                Little Portugal, Trinity   
Yoga Studio                                       University of Toronto, Harbord   

                               Neighborhood Latitude  Neighborhood Longitude  \
Venue Category                                                                 
Adult Boutique                             43.665860              -79.383160   
Airport                                    43.628947              -79.394420   
Airport Food Court                         43.628947              -79.394420   
Airport Lounge                             43.628947              -79.394420   
Airport Service                            43.628947              -79.394420   
...                                              ...                     ...   
Vegetarian / Vegan Restaurant              43.657952              -79.373306   
Video Game Store                           43.662696              -79.400049   
Vietnamese Restaurant                      43.686412              -79.400049   
Wine Bar                                   43.653206              -79.400049   
Yoga Studio                                43.715383              -79.321558   

                                                                           Venue  \
Venue Category                                                                     
Adult Boutique                                                         Seduction   
Airport                        Billy Bishop Toronto City Airport (YTZ) (Billy...   
Airport Food Court                                             Billy Bishop Café   
Airport Lounge                                                     Porter Lounge   
Airport Service                                         ORNGE - Toronto Air Base   
...                                                                          ...   
Vegetarian / Vegan Restaurant                                             Veghed   
Video Game Store                                                     A & C Games   
Vietnamese Restaurant                                                     Sprout   
Wine Bar                                                         Paris Paris Bar   
Yoga Studio                                                            YogaSpace   

                               Venue Latitude  Venue Longitude  
Venue Category                                                  
Adult Boutique                      43.665620       -79.384681  
Airport                             43.631683       -79.396033  
Airport Food Court                  43.631132       -79.396139  
Airport Lounge                      43.631360       -79.395756  
Airport Service                     43.631662       -79.398166  
...                                       ...              ...  
Vegetarian / Vegan Restaurant       43.656016       -79.374453  
Video Game Store                    43.664939       -79.403194  
Vietnamese Restaurant               43.687996       -79.394651  
Wine Bar                            43.653479       -79.401427  
Yoga Studio                         43.714070       -79.32433

#### There are 198 Venue Categories

In [45]:
venues_in_toronto.groupby("Venue Category").count()

Neighborhood  Neighborhood Latitude  \
Venue Category                                                       
Adult Boutique                            1                      1   
Airport                                   1                      1   
Airport Food Court                        1                      1   
Airport Lounge                            2                      2   
Airport Service                           1                      1   
...                                     ...                    ...   
Vegetarian / Vegan Restaurant            10                     10   
Video Game Store                          1                      1   
Vietnamese Restaurant                     5                      5   
Wine Bar                                  2                      2   
Yoga Studio                               9                      9   

                               Neighborhood Longitude  Venue  Venue Latitude  \
Venue Category                                                                 
Adult Boutique                                      1      1               1   
Airport                                             1      1               1   
Airport Food Court                                  1      1               1   
Airport Lounge                                      2      2               2   
Airport Service                                     1      1               1   
...                                               ...    ...             ...   
Vegetarian / Vegan Restaurant                      10     10              10   
Video Game Store                                    1      1               1   
Vietnamese Restaurant                               5      5               5   
Wine Bar                                            2      2               2   
Yoga Studio                                         9      9               9   

                               Venue Longitude  
Venue Category                                  
Adult Boutique                               1  
Airport                                      1  
Airport Food Court                           1  
Airport Lounge                               2  
Airport Service                              1  
...                                        ...  
Vegetarian / Vegan Restaurant               10  
Video Game Store                             1  
Vietnamese Restaurant                        5  
Wine Bar                                     2  
Yoga Studio                                  9  

[198 rows x 6 columns]

### Analysing the Toronto Neighborhood

#### We find that the 'Venue Category' column of our "venues_in_toronto_dfnew" and "venues_in_toronto" Dataframes have values named "Neighborhood"

In [46]:
venues_in_toronto_dfnew[venues_in_toronto_dfnew["Venue Category"] == "Neighborhood"]

Neighborhood  Neighborhood Latitude  \
269                                        The Beaches              43.676357   
381                                    Studio District              43.659526   
709                           Richmond, Adelaide, King              43.650571   
730  Harbourfront East, Union Station, Toronto Islands              43.640816   

     Neighborhood Longitude             Venue  Venue Latitude  \
269              -79.293031     Upper Beaches       43.680563   
381              -79.340923       Leslieville       43.662070   
709              -79.384568  Downtown Toronto       43.653232   
730              -79.381752      Harbourfront       43.639526   

     Venue Longitude Venue Category  
269       -79.292869   Neighborhood  
381       -79.337856   Neighborhood  
709       -79.385296   Neighborhood  
730       -79.380688   Neighborhood

In [47]:
venues_in_toronto[venues_in_toronto["Venue Category"] == "Neighborhood"]

Neighborhood  Neighborhood Latitude  \
3                                          The Beaches              43.676357   
65                                     Studio District              43.659526   
393                           Richmond, Adelaide, King              43.650571   
414  Harbourfront East, Union Station, Toronto Islands              43.640816   

     Neighborhood Longitude             Venue  Venue Latitude  \
3                -79.293031     Upper Beaches       43.680563   
65               -79.340923       Leslieville       43.662070   
393              -79.384568  Downtown Toronto       43.653232   
414              -79.381752      Harbourfront       43.639526   

     Venue Longitude Venue Category  
3         -79.292869   Neighborhood  
65        -79.337856   Neighborhood  
393       -79.385296   Neighborhood  
414       -79.380688   Neighborhood

#### We need to change it to something else, like "Neighbourhood", so that it does not clash with the "Neighborhood" column from the Main Dataframe

In [48]:
venues_in_toronto_dfnew.replace("Neighborhood", "Neighbourhood", inplace = True)
venues_in_toronto_dfnew[venues_in_toronto_dfnew["Venue Category"] == "Neighbourhood"]

Neighborhood  Neighborhood Latitude  \
269                                        The Beaches              43.676357   
381                                    Studio District              43.659526   
709                           Richmond, Adelaide, King              43.650571   
730  Harbourfront East, Union Station, Toronto Islands              43.640816   

     Neighborhood Longitude             Venue  Venue Latitude  \
269              -79.293031     Upper Beaches       43.680563   
381              -79.340923       Leslieville       43.662070   
709              -79.384568  Downtown Toronto       43.653232   
730              -79.381752      Harbourfront       43.639526   

     Venue Longitude Venue Category  
269       -79.292869  Neighbourhood  
381       -79.337856  Neighbourhood  
709       -79.385296  Neighbourhood  
730       -79.380688  Neighbourhood

In [49]:
venues_in_toronto.replace("Neighborhood", "Neighbourhood", inplace = True)
venues_in_toronto[venues_in_toronto["Venue Category"] == "Neighbourhood"]

Neighborhood  Neighborhood Latitude  \
3                                          The Beaches              43.676357   
65                                     Studio District              43.659526   
393                           Richmond, Adelaide, King              43.650571   
414  Harbourfront East, Union Station, Toronto Islands              43.640816   

     Neighborhood Longitude             Venue  Venue Latitude  \
3                -79.293031     Upper Beaches       43.680563   
65               -79.340923       Leslieville       43.662070   
393              -79.384568  Downtown Toronto       43.653232   
414              -79.381752      Harbourfront       43.639526   

     Venue Longitude Venue Category  
3         -79.292869  Neighbourhood  
65        -79.337856  Neighbourhood  
393       -79.385296  Neighbourhood  
414       -79.380688  Neighbourhood

#### Now we can proceed with One Hot Encoding

In [50]:
toronto_ohe_dfnew = pd.get_dummies(venues_in_toronto_dfnew[['Venue Category']], prefix="", prefix_sep="")
toronto_ohe_dfnew.head()

Accessories Store  Adult Boutique  Airport  Airport Food Court  \
0                  0               0        0                   0   
1                  0               0        0                   0   
2                  0               0        0                   0   
3                  0               0        0                   0   
4                  0               0        0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  ...  Truck Stop  Vegetarian / Vegan Restaurant  \
0             0         0  ...           0                              0   
1             0         0  ...           0                              0   
2             0         0  ...           0                              0   
3             0         0  ...           0                              0   
4             0         0  ...           0                              0   

   Video Game Store  Video Store  Vietnamese Restaurant  Warehouse Store  \
0                 0            0                      0                0   
1                 0            0                      0                0   
2                 0            0                      0                0   
3                 0            0                      0                0   
4                 0            0                      0                0   

   Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0         0            0              0            0  
1         0            0              0            0  
2         0            0              0            0  
3         0            0              0            0  
4         0            0              0            0  

[5 rows x 240 columns]

In [51]:
toronto_ohe = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_ohe.head()

Adult Boutique  Airport  Airport Food Court  Airport Lounge  \
0               0        0                   0               0   
1               0        0                   0               0   
2               0        0                   0               0   
3               0        0                   0               0   
4               0        0                   0               0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                0                 0                    0             0   
1                0                 0                    0             0   
2                0                 0                    0             0   
3                0                 0                    0             0   
4                0                 0                    0             0   

   Aquarium  Art Gallery  ...  Theme Restaurant  Tibetan Restaurant  \
0         0            0  ...                 0                   0   
1         0            0  ...                 0                   0   
2         0            0  ...                 0                   0   
3         0            0  ...                 0                   0   
4         0            0  ...                 0                   0   

   Toy / Game Store  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0                 0      1              0                              0   
1                 0      0              0                              0   
2                 0      0              0                              0   
3                 0      0              0                              0   
4                 0      0              0                              0   

   Video Game Store  Vietnamese Restaurant  Wine Bar  Yoga Studio  
0                 0                      0         0            0  
1                 0                      0         0            0  
2                 0                      0         0            0  
3                 0                      0         0            0  
4                 0                      0         0            0  

[5 rows x 198 columns]

#### Now we add the "Neighborhood" Column from the Main Dataframe to the Encoded Dataframe

In [52]:
toronto_ohe_dfnew["Neighborhood"] = venues_in_toronto_dfnew["Neighborhood"] # This adds the "Neighborhood" column in the end

# Moving the Neighborhood Column to the First Column
columns = [toronto_ohe_dfnew.columns[-1]] + list(toronto_ohe_dfnew.columns[:-1])
toronto_ohe_dfnew = toronto_ohe_dfnew[columns]

toronto_ohe_dfnew.head()

Neighborhood  Accessories Store  Adult Boutique  \
0                          Malvern, Rouge                  0               0   
1  Rouge Hill, Port Union, Highland Creek                  0               0   
2       Guildwood, Morningside, West Hill                  0               0   
3       Guildwood, Morningside, West Hill                  0               0   
4       Guildwood, Morningside, West Hill                  0               0   

   Airport  Airport Food Court  Airport Lounge  Airport Service  \
0        0                   0               0                0   
1        0                   0               0                0   
2        0                   0               0                0   
3        0                   0               0                0   
4        0                   0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  ...  Truck Stop  \
0                 0                    0             0  ...           0   
1                 0                    0             0  ...           0   
2                 0                    0             0  ...           0   
3                 0                    0             0  ...           0   
4                 0                    0             0  ...           0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 241 columns]

In [53]:
toronto_ohe["Neighborhood"] = venues_in_toronto["Neighborhood"] # This adds the "Neighborhood" column in the end

# Moving the Neighborhood Column to the First Column
columns = [toronto_ohe.columns[-1]] + list(toronto_ohe.columns[:-1])
toronto_ohe = toronto_ohe[columns]

toronto_ohe.head()

Neighborhood  Adult Boutique  Airport  Airport Food Court  \
0         The Beaches               0        0                   0   
1         The Beaches               0        0                   0   
2         The Beaches               0        0                   0   
3         The Beaches               0        0                   0   
4  The Danforth  East               0        0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  ...  Theme Restaurant  Tibetan Restaurant  \
0             0         0  ...                 0                   0   
1             0         0  ...                 0                   0   
2             0         0  ...                 0                   0   
3             0         0  ...                 0                   0   
4             0         0  ...                 0                   0   

   Toy / Game Store  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0                 0      1              0                              0   
1                 0      0              0                              0   
2                 0      0              0                              0   
3                 0      0              0                              0   
4                 0      0              0                              0   

   Video Game Store  Vietnamese Restaurant  Wine Bar  Yoga Studio  
0                 0                      0         0            0  
1                 0                      0         0            0  
2                 0                      0         0            0  
3                 0                      0         0            0  
4                 0                      0         0            0  

[5 rows x 199 columns]

#### Next, let's group rows by Neighborhood and by taking the mean of the frequency of occurrence of each Venue Category

In [54]:
toronto_grouped_dfnew = toronto_ohe_dfnew.groupby('Neighborhood').mean().reset_index()
toronto_grouped_dfnew

Neighborhood  Accessories Store  \
0                                         Agincourt                0.0   
1                            Alderwood, Long Branch                0.0   
2   Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                   Bayview Village                0.0   
4                 Bedford Park, Lawrence Manor East                0.0   
..                                              ...                ...   
95                                  Willowdale West                0.0   
96                          Willowdale, Newtonbrook                0.0   
97                                           Woburn                0.0   
98                                 Woodbine Heights                0.0   
99                                  York Mills West                0.0   

    Adult Boutique  Airport  Airport Food Court  Airport Lounge  \
0              0.0      0.0                 0.0             0.0   
1              0.0      0.0                 0.0             0.0   
2              0.0      0.0                 0.0             0.0   
3              0.0      0.0                 0.0             0.0   
4              0.0      0.0                 0.0             0.0   
..             ...      ...                 ...             ...   
95             0.0      0.0                 0.0             0.0   
96             0.0      0.0                 0.0             0.0   
97             0.0      0.0                 0.0             0.0   
98             0.0      0.0                 0.0             0.0   
99             0.0      0.0                 0.0             0.0   

    Airport Service  Airport Terminal  American Restaurant  Antique Shop  ...  \
0               0.0               0.0                 0.00           0.0  ...   
1               0.0               0.0                 0.00           0.0  ...   
2               0.0               0.0                 0.00           0.0  ...   
3               0.0               0.0                 0.00           0.0  ...   
4               0.0               0.0                 0.04           0.0  ...   
..              ...               ...                  ...           ...  ...   
95              0.0               0.0                 0.00           0.0  ...   
96              0.0               0.0                 0.00           0.0  ...   
97              0.0               0.0                 0.00           0.0  ...   
98              0.0               0.0                 0.00           0.0  ...   
99              0.0               0.0                 0.00           0.0  ...   

    Truck Stop  Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0          0.0                            0.0               0.0        0.000   
1          0.0                            0.0               0.0        0.000   
2          0.0                            0.0               0.0        0.000   
3          0.0                            0.0               0.0        0.000   
4          0.0                            0.0               0.0        0.000   
..         ...                            ...               ...          ...   
95         0.0                            0.0               0.0        0.000   
96         0.0                            0.0               0.0        0.000   
97         0.0                            0.0               0.0        0.000   
98         0.0                            0.0               0.0        0.125   
99         0.0                            0.0               0.0        0.000   

    Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                     0.0              0.0       0.0          0.0   
1                     0.0              0.0       0.0          0.0   
2                     0.0              0.0       0.0          0.0   
3                     0.0              0.0       0.0          0.0   
4                     0.0              0.0       0.0          0.0 

In [55]:
toronto_grouped = toronto_ohe.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Adult Boutique  \
0                                         Berczy Park        0.000000   
1        Brockton, Parkdale Village, Exhibition Place        0.000000   
2   CN Tower, King and Spadina, Railway Lands, Har...        0.000000   
3                                  Central Bay Street        0.000000   
4                                            Christie        0.000000   
5                                Church and Wellesley        0.033333   
6                      Commerce Court, Victoria Hotel        0.000000   
7                                          Davisville        0.000000   
8                                    Davisville North        0.000000   
9                        Dufferin, Dovercourt Village        0.000000   
10                                     Enclave of M4L        0.000000   
11                                     Enclave of M5E        0.000000   
12             First Canadian Place, Underground city        0.000000   
13                           Forest Hill North & West        0.000000   
14                           Garden District, Ryerson        0.000000   
15  Harbourfront East, Union Station, Toronto Islands        0.000000   
16                      High Park, The Junction South        0.000000   
17                     India Bazaar, The Beaches West        0.000000   
18          Kensington Market, Chinatown, Grange Park        0.000000   
19                                      Lawrence Park        0.000000   
20                           Little Portugal, Trinity        0.000000   
21                        Moore Park, Summerhill East        0.000000   
22                                 North Toronto West        0.000000   
23                             Parkdale, Roncesvalles        0.000000   
24                          Regent Park, Harbourfront        0.000000   
25                           Richmond, Adelaide, King        0.000000   
26                                           Rosedale        0.000000   
27                                           Roselawn        0.000000   
28                                 Runnymede, Swansea        0.000000   
29                                     St. James Town        0.000000   
30                        St. James Town, Cabbagetown        0.000000   
31                                    Studio District        0.000000   
32  Summerhill West, Rathnelly, South Hill, Forest...        0.000000   
33                The Annex, North Midtown, Yorkville        0.000000   
34                                        The Beaches        0.000000   
35                                 The Danforth  East        0.000000   
36                       The Danforth West, Riverdale        0.000000   
37           Toronto Dominion Centre, Design Exchange        0.000000   
38                     University of Toronto, Harbord        0.000000   

     Airport  Airport Food Court  Airport Lounge  Airport Service  \
0   0.000000            0.000000        0.000000         0.000000   
1   0.000000            0.000000        0.000000         0.000000   
2   0.071429            0.071429        0.142857         0.071429   
3   0.000000            0.000000        0.000000         0.000000   
4   0.000000            0.000000        0.000000         0.000000   
5   0.000000            0.000000        0.000000         0.000000   
6   0.000000            0.000000        0.000000         0.000000   
7   0.000000            0.000000        0.000000         0.000000   
8   0.000000            0.000000        0.000000         0.000000   
9   0.000000            0.000000        0.000000         0.000000   
10  0.000000            0.000000        0.000000         0.000000   
11  0.000000            0.000000        0.000000         0.000000   
12  0.000000            0.000000        0.000000         0.000000   
13  0.000000            0.000000        0.000000         0.000000   
14  0.000000            0.000000        0.000000         0.000000   
15  0.000000    

### Note: From this point forward, we will do the Analysis only on the "toronto_grouped_dfnew" Dataframe

In [56]:
toronto_grouped_dfnew.shape

(100, 241)

#### Print each Neighborhood along with the top 5 most common Venues

In [57]:
num_top_venues = 5

for nhood in toronto_grouped_dfnew['Neighborhood']:
    print("---------"+nhood+"---------")
    temp = toronto_grouped_dfnew[toronto_grouped_dfnew['Neighborhood'] == nhood].T.reset_index()
    temp.columns = ["Venue", "Frequency"]
    temp = temp.iloc[1:]
    temp["Frequency"] = temp["Frequency"].astype(float)
    temp = temp.round({"Frequency" : 2})
    print(temp.sort_values("Frequency", ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

---------Agincourt---------
                       Venue  Frequency
0                     Lounge       0.25
1             Breakfast Spot       0.25
2             Clothing Store       0.25
3  Latin American Restaurant       0.25
4          Accessories Store       0.00


---------Alderwood, Long Branch---------
            Venue  Frequency
0     Pizza Place       0.25
1            Pool       0.12
2  Sandwich Place       0.12
3    Dance Studio       0.12
4     Coffee Shop       0.12


---------Bathurst Manor, Wilson Heights, Downsview North---------
                       Venue  Frequency
0                       Bank       0.10
1                Coffee Shop       0.10
2                Pizza Place       0.05
3                Gas Station       0.05
4  Middle Eastern Restaurant       0.05


---------Bayview Village---------
                 Venue  Frequency
0  Japanese Restaurant       0.25
1                 Café       0.25
2   Chinese Restaurant       0.25
3                 Bank       0.25
4

                Venue  Frequency
0      Discount Store        0.2
1         Bus Station        0.2
2    Department Store        0.2
3         Coffee Shop        0.2
4  Chinese Restaurant        0.2


---------Kensington Market, Chinatown, Grange Park---------
                   Venue  Frequency
0                   Café       0.10
1   Caribbean Restaurant       0.07
2  Vietnamese Restaurant       0.07
3                 Bakery       0.07
4            Fish Market       0.03


---------Kingsview Village, St. Phillips, Martin Grove Gardens, Richview Gardens---------
               Venue  Frequency
0               Park       0.33
1  Mobile Phone Shop       0.33
2     Sandwich Place       0.33
3  Accessories Store       0.00
4      Movie Theater       0.00


---------Lawrence Manor, Lawrence Heights---------
                    Venue  Frequency
0          Clothing Store       0.45
1       Accessories Store       0.09
2   Vietnamese Restaurant       0.09
3             Event Space       0.09
4 

               Venue  Frequency
0               Café       0.14
1     Sandwich Place       0.14
2        Coffee Shop       0.10
3        Pizza Place       0.05
4  Indian Restaurant       0.05


---------The Beaches---------
               Venue  Frequency
0      Neighbourhood       0.25
1  Health Food Store       0.25
2              Trail       0.25
3                Pub       0.25
4        Men's Store       0.00


---------The Danforth  East---------
                 Venue  Frequency
0         Intersection       0.25
1    Convenience Store       0.25
2                 Park       0.25
3  Rental Car Location       0.25
4               Museum       0.00


---------The Danforth West, Riverdale---------
                Venue  Frequency
0    Greek Restaurant       0.23
1  Italian Restaurant       0.07
2      Ice Cream Shop       0.07
3         Yoga Studio       0.03
4     Bubble Tea Shop       0.03


---------The Kingsway, Montgomery Road, Old Mill North---------
               Venue  Freque

### Let's put that into a Pandas Dataframe

#### Write a function to sort the Venues in descending order

In [58]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

#### Create the new dataframe and display the top 10 Venues for each Neighborhood

In [59]:
num_top_venues = 10

indicators = ["st", "nd", "rd"]

# Create columns according to number of top Venues
columns = ["Neighborhood"]
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# Create a new Dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns = columns)
neighborhoods_venues_sorted["Neighborhood"] = toronto_grouped_dfnew["Neighborhood"]

for ind in np.arange(toronto_grouped_dfnew.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped_dfnew.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                  Bank   
3                                  Bayview Village   Japanese Restaurant   
4                Bedford Park, Lawrence Manor East    Italian Restaurant   

  2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
0        Breakfast Spot        Clothing Store  Latin American Restaurant   
1                  Pool        Sandwich Place               Dance Studio   
2           Coffee Shop           Pizza Place                Gas Station   
3                  Café    Chinese Restaurant                       Bank   
4           Coffee Shop        Sandwich Place    Comfort Food Restaurant   

       5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0          Accessories Store                Museum           Men's Store   
1                Coffee Shop                   Pub                   Gym   
2  Middle Eastern Restaurant              Pharmacy     Mobile Phone Shop   
3                Men's Store         Metro Station    Mexican Restaurant   
4               Cupcake Shop                   Pub            Restaurant   

       8th Most Common Venue 9th Most Common Venue     10th Most Common Venue  
0              Metro Station    Mexican Restaurant  Middle Eastern Restaurant  
1                Men's Store         Metro Station         Mexican Restaurant  
2              Shopping Mall           Supermarket           Sushi Restaurant  
3  Middle Eastern Restaurant    Miscellaneous Shop          Mobile Phone Shop  
4                       Café               Butcher       Fast Food Restaurant

### Build a Model to Cluster Neighborhood

#### Run k-means to cluster the Neighborhood into 5 clusters

In [60]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [61]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped_dfnew.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters = k_num_clusters, random_state = 0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

In [62]:
# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:100]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 3, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0,
       4, 0, 0, 0, 0, 3, 0, 3, 0, 0, 0, 3, 0, 3, 0, 3, 0, 0, 4, 0, 0, 0,
       3, 0, 0, 3, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 3, 0, 0,
       0, 0, 3, 0, 1, 0, 0, 0, 1, 0, 0, 3])

#### Create a new Dataframe that includes the Cluster as well as the Top 10 Venues for each Neighborhood

In [63]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

#### Join "toronto_grouped_dfnew" with "df_new" on Neighborhood" to add Latitude & Longitude for each Neighborhood to prepare it for plotting

In [64]:
toronto_merged = df_new

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index("Neighborhood"), on = "Neighborhood")

toronto_merged.head()

Postal Code      Borough                            Neighborhood   Latitude  \
0         M1B  Scarborough                          Malvern, Rouge  43.806686   
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.194353             0.0  Fast Food Restaurant     Accessories Store   
1 -79.160497             2.0                   Bar     Accessories Store   
2 -79.188711             0.0        Medical Center     Electronics Store   
3 -79.216917             0.0           Coffee Shop          Soccer Field   
4 -79.239476             0.0  Caribbean Restaurant                Bakery   

   3rd Most Common Venue 4th Most Common Venue     5th Most Common Venue  \
0    Martial Arts School        Medical Center  Mediterranean Restaurant   
1    Martial Arts School        Medical Center  Mediterranean Restaurant   
2                   Bank          Intersection                Donut Shop   
3  Korean BBQ Restaurant     Accessories Store                    Museum   
4                   Bank           Gas Station        Athletics & Sports   

      6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0               Men's Store         Metro Station    Mexican Restaurant   
1               Men's Store         Metro Station    Mexican Restaurant   
2                Restaurant   Rental Car Location    Mexican Restaurant   
3  Mediterranean Restaurant           Men's Store         Metro Station   
4       Fried Chicken Joint       Thai Restaurant      Hakka Restaurant   

       9th Most Common Venue     10th Most Common Venue  
0  Middle Eastern Restaurant         Miscellaneous Shop  
1  Middle Eastern Restaurant         Miscellaneous Shop  
2             Breakfast Spot  Middle Eastern Restaurant  
3         Mexican Restaurant  Middle Eastern Restaurant  
4         Miscellaneous Shop          Mobile Phone Shop

#### Drop all the NaN values to avoid skewing of data

In [65]:
toronto_merged = toronto_merged.dropna(subset = ["Cluster Labels"])

#### Visualise the resulting Clusters

In [66]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [67]:
# Create Map
map_clusters = folium.Map(location = [latitude, longitude], zoom_start = 11)

# Set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i * x) ** 2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# Add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged["Latitude"], toronto_merged["Longitude"], toronto_merged["Neighborhood"], toronto_merged["Cluster Labels"]):
    label = folium.Popup('Cluster ' + str(int(cluster) + 1) + '\n' + str(poi) , parse_html = True)
    folium.CircleMarker(
                            [lat, lon],
                            radius = 5,
                            popup = label,
                            color = rainbow[int(cluster-1)],
                            fill = True,
                            fill_color = rainbow[int(cluster-1)]
                        ).add_to(map_clusters)
        
map_clusters

### Examine Clusters

#### Cluster 1:

In [68]:
toronto_merged.loc[toronto_merged["Cluster Labels"] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0           Scarborough             0.0  Fast Food Restaurant   
2           Scarborough             0.0        Medical Center   
3           Scarborough             0.0           Coffee Shop   
4           Scarborough             0.0  Caribbean Restaurant   
5           Scarborough             0.0            Playground   
..                  ...             ...                   ...   
95            Etobicoke             0.0           Pizza Place   
96           North York             0.0           Pizza Place   
99            Etobicoke             0.0           Pizza Place   
101           Etobicoke             0.0         Grocery Store   
102  EtobicokeNorthwest             0.0             Drugstore   

    2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0       Accessories Store        Martial Arts School        Medical Center   
2       Electronics Store                       Bank          Intersection   
3            Soccer Field      Korean BBQ Restaurant     Accessories Store   
4                  Bakery                       Bank           Gas Station   
5          Adult Boutique             Massage Studio        Medical Center   
..                    ...                        ...                   ...   
95      Convenience Store                   Pharmacy             Pet Store   
96         Clothing Store                 Food Truck     Accessories Store   
99             Playground  Middle Eastern Restaurant        Discount Store   
101  Fast Food Restaurant        Fried Chicken Joint              Pharmacy   
102   Rental Car Location                        Bar            Truck Stop   

        5th Most Common Venue       6th Most Common Venue  \
0    Mediterranean Restaurant                 Men's Store   
2                  Donut Shop                  Restaurant   
3                      Museum    Mediterranean Restaurant   
4          Athletics & Sports         Fried Chicken Joint   
5    Mediterranean Restaurant                 Men's Store   
..                        ...                         ...   
95                Coffee Shop                        Café   
96              Movie Theater              Medical Center   
99               Intersection          Chinese Restaurant   
101               Coffee Shop              Sandwich Place   
102       Monument / Landmark  Modern European Restaurant   

        7th Most Common Venue 8th Most Common Venue  \
0               Metro Station    Mexican Restaurant   
2         Rental Car Location    Mexican Restaurant   
3                 Men's Store         Metro Station   
4             Thai Restaurant      Hakka Restaurant   
5               Metro Station    Mexican Restaurant   
..                        ...                   ...   
95             Shopping Plaza            Beer Store   
96   Mediterranean Restaurant           Men's Store   
99             Sandwich Place           Coffee Shop   
101                Beer Store           Pizza Place   
102         Mobile Phone Shop    Miscellaneous Shop   

         9th Most Common Venue     10th Most Common Venue  
0    Middle Eastern Restaurant         Miscellaneous Shop  
2               Breakfast Spot  Middle Eastern Restaurant  
3           Mexican Restaurant  Middle Eastern Restaurant  
4           Miscellaneous Shop          Mobile Phone Shop  
5    Middle Eastern Restaurant         Miscellaneous Shop  
..                         ...                        ...  
95                Liquor Store          Accessories Store  
96               Metro Station         Mexican Restaurant  
99          Mexican Restaurant         Miscellaneous Shop  
101               Home Service                Music Venue  
102          Accessories Store              Movie Theater  

[80 rows x 12 columns]

#### Cluster 2:

In [69]:
toronto_merged.loc[toronto_merged["Cluster Labels"] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
21  North York             1.0                  Park     Accessories Store   
98        York             1.0                  Park     Accessories Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
21         Movie Theater        Massage Studio        Medical Center   
98         Movie Theater        Massage Studio        Medical Center   

       6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
21  Mediterranean Restaurant           Men's Store         Metro Station   
98  Mediterranean Restaurant           Men's Store         Metro Station   

   9th Most Common Venue     10th Most Common Venue  
21    Mexican Restaurant  Middle Eastern Restaurant  
98    Mexican Restaurant  Middle Eastern Restaurant

#### Cluster 3:

In [70]:
toronto_merged.loc[toronto_merged["Cluster Labels"] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
1  Scarborough             2.0                   Bar     Accessories Store   

  3rd Most Common Venue 4th Most Common Venue     5th Most Common Venue  \
1   Martial Arts School        Medical Center  Mediterranean Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
1           Men's Store         Metro Station    Mexican Restaurant   

       9th Most Common Venue 10th Most Common Venue  
1  Middle Eastern Restaurant     Miscellaneous Shop

#### Cluter 4:

In [71]:
toronto_merged.loc[toronto_merged["Cluster Labels"] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels       1st Most Common Venue  \
7              Scarborough             3.0                      Bakery   
14             Scarborough             3.0                  Playground   
23              North York             3.0  Construction & Landscaping   
25              North York             3.0        Fast Food Restaurant   
30              North York             3.0                     Airport   
31              North York             3.0               Grocery Store   
40   East YorkEast Toronto             3.0                Intersection   
44         Central Toronto             3.0          Dim Sum Restaurant   
48         Central Toronto             3.0                 Summer Camp   
50        Downtown Toronto             3.0                        Park   
74                    York             3.0                        Park   
79              North York             3.0  Construction & Landscaping   
90               Etobicoke             3.0                        Pool   
94               Etobicoke             3.0                      Bakery   
100              Etobicoke             3.0                        Park   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
7                Bus Line         Metro Station          Soccer Field   
14           Intersection                  Park        Medical Center   
23      Convenience Store                  Park     Accessories Store   
25                   Park     Food & Drink Shop     Accessories Store   
30                   Park      Business Service     Accessories Store   
31                   Park                  Bank         Shopping Mall   
40      Convenience Store                  Park   Rental Car Location   
44                   Park                  Lake           Swim School   
48                   Park          Tennis Court     Accessories Store   
50             Playground                 Trail         Movie Theater   
74          Women's Store                  Pool    Mexican Restaurant   
79                   Park      Basketball Court                Bakery   
90                   Park            Smoke Shop                 River   
94      Accessories Store   Martial Arts School        Medical Center   
100     Mobile Phone Shop        Sandwich Place     Accessories Store   

        5th Most Common Venue       6th Most Common Venue  \
7                Intersection                        Park   
14   Mediterranean Restaurant                 Men's Store   
23              Movie Theater              Medical Center   
25                      Motel              Massage Studio   
30              Movie Theater              Medical Center   
31          Accessories Store                       Motel   
40                     Museum    Mediterranean Restaurant   
44                   Bus Line               Movie Theater   
48                      Motel              Massage Studio   
50             Massage Studio              Medical Center   
74        Monument / Landmark  Modern European Restaurant   
79          Accessories Store               Movie Theater   
90          Accessories Store                       Motel   
94   Mediterranean Restaurant                 Men's Store   
100             Movie Theater              Massage Studio   

        7th Most Common Venue     8th Most Common Venue  \
7                 Bus Station            Ice Cream Shop   
14              Metro Station        Mexican Restaurant   
23   Mediterranean Restaurant               Men's Store   
25             Medical Center  Mediterranean Restaurant   
30   Mediterranean Restaurant               Men's Store   
31             Medical Center  Mediterranean Restaurant   
40                Men's Store             Metro Station   
44   Mediterranean Restaurant               Men's Store   
48             Medical Center  Mediterranean Restaurant   
50   Mediterranean Restaurant               Men's Store   
74          Mobile Phone Shop        Mi

#### Cluster 5:

In [72]:
toronto_merged.loc[toronto_merged["Cluster Labels"] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
91   Etobicoke             4.0        Baseball Field     Accessories Store   
97  North York             4.0        Baseball Field     Accessories Store   

   3rd Most Common Venue 4th Most Common Venue     5th Most Common Venue  \
91   Martial Arts School        Medical Center  Mediterranean Restaurant   
97   Martial Arts School        Medical Center  Mediterranean Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
91           Men's Store         Metro Station    Mexican Restaurant   
97           Men's Store         Metro Station    Mexican Restaurant   

        9th Most Common Venue 10th Most Common Venue  
91  Middle Eastern Restaurant     Miscellaneous Shop  
97  Middle Eastern Restaurant     Miscellaneous Shop